### PR TITLE
Features/optimizations

### PR DESCRIPTION
**Pull request by brason**
Various additions from three.js and some optimizations. I'll add some tests tomorrow.

### CODE REVIEW

File: lib/src/vector_math/quaternion.dart [hunk 1]
```diff
@@ -150,6 +150,24 @@ class Quaternion {
     }
   }
 
+  static final Matrix3 _m3 = new Matrix3.zero();
+
+  /// Set the quaternion with rotation from the upper 3x3 of [arg].
+  void setFromRotation4(Matrix4 arg) {
+    final argStorage = arg._m4storage;
+    final m3Storage = _m3._m3storage;
+    m3Storage[0] = argStorage[0];
```
Comment (Fox32) [inline lib/src/vector_math/quaternion.dart]: There is a Matrix4.copyRotation that you could use here

**Comment by Fox32** [comment]
No tests for the new features?

**Comment by brason** [comment]
I'll do it tomorrow, It's late. :)

File: lib/src/vector_math/quaternion.dart (hunk 1 above)
Comment (brason) [inline lib/src/vector_math/quaternion.dart]: Changed it, thanks.

**Comment by brason** [comment]
@kevmoo Why did you close it?

**Comment by kevmoo** [comment]
Oops! Total mistake on my party.

**Comment by brason** [comment]
Haha :-) So do I have to open a new PR?

**Comment by kevmoo** [comment]
It looks like it. I can't re-open a pull request.

My apologies!

**Comment by brason** [comment]
It's all cool ;-)

**Pull request closed without merging**